framework/src/preconditioners/VariableCondensationPreconditioner.C

```diff
@@ -0,0 +1,712 @@
+                   ".");
+      for (auto i : index_range(di))
+      {
+        // when we have adaptive condensation, skip when di does not contain any indices in
+        // _zero_rows
+        if (std::find(_zero_rows.begin(), _zero_rows.end(), di[i]) == _zero_rows.end() &&
+            _adaptive_condensation)
+          break;
+        _glm.push_back(di[i]);
+        if (_dofmap->local_index(di[i]))
+          _lm.push_back(di[i]);
+
+        // save the corresponding coupled dof indices
+        _gu2c.push_back(cp_di[i]);
+        if (_dofmap->local_index(cp_di[i]))
+          _u2c.push_back(cp_di[i]);
+        _map_glm_gu2c.insert(std::make_pair(di[i], cp_di[i]));
+        _map_gu2c_glm.insert(std::make_pair(cp_di[i], di[i]));
+      }
+    }
+
+  // check if we endup with none dof to condense
+  if (_glm.empty())
+  {
+    _need_condense = false;
+#ifdef DEBUG
+    mooseWarning("The variable provided do not have a saddle-point character. VCP will "
+                 "continue without condensing the dofs.");
+#endif
+    return;
+  }
+  else
+    _need_condense = true;
+
+  std::sort(_glm.begin(), _glm.end());
+  std::sort(_lm.begin(), _lm.end());
+
+  std::sort(_gu2c.begin(), _gu2c.end());
+  std::sort(_u2c.begin(), _u2c.end());
+
+  for (auto i : index_range(_glm))
+    _map_gu2c_to_order.insert(std::make_pair(_map_glm_gu2c.at(_glm[i]), i));
+}
+
+void
+VariableCondensationPreconditioner::getDofColRow()
+{
+  // clean the containers if we want to update the dofs
+  if (!_grows.empty())
+    _grows.clear();
+  if (!_rows.empty())
+    _rows.clear();
+  if (!_gcols.empty())
+    _gcols.clear();
+  if (!_cols.empty())
+    _cols.clear();
+  // row: all without u2c
+  // col: all without lm
+  for (numeric_index_type i = 0; i < _dofmap->n_dofs(); ++i)
+  {
+    if (std::find(_gu2c.begin(), _gu2c.end(), i) != _gu2c.end())
+      continue;
+
+    _grows.push_back(i);
+    if (_dofmap->local_index(i))
+      _rows.push_back(i);
+
+    // ensure the lm and u2c correspondance
+    if (_map_glm_gu2c.find(i) != _map_glm_gu2c.end())
+    {
+      auto u2c_idx = _map_glm_gu2c.at(i);
+      _gcols.push_back(u2c_idx);
+
+      if (_dofmap->local_index(u2c_idx))
+        _cols.push_back(u2c_idx);
+    }
+    else
+    {
+      _gcols.push_back(i);
+
+      if (_dofmap->local_index(i))
+        _cols.push_back(i);
+    }
+  }
+}
+
+void
+VariableCondensationPreconditioner::init()
+{
+  TIME_SECTION(_init_timer);
+
+  if (!_preconditioner)
+    _preconditioner =
+        Preconditioner<Number>::build_preconditioner(MoosePreconditioner::_communicator);
+
+  _is_initialized = true;
+}
+
+void
+VariableCondensationPreconditioner::condenseSystem()
+{
+  // extract _M from the original matrix
+  _matrix->create_submatrix(*_M, _rows, _lm);
+
+  // get the row associated with u2c
+  _u2c_rows->init(_gu2c.size(), _gcols.size(), _u2c.size(), _cols.size());
+  MatSetOption(_u2c_rows->mat(), MAT_NEW_NONZERO_ALLOCATION_ERR, PETSC_FALSE);
```

Comment:
We may need to fix this. The code will be very slow if allocation is not right

Reply:
Yeah, glad that you point this out. I have no idea how to improve these allocations.. any suggestions?